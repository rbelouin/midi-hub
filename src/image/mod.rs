#[derive(Copy, Clone, Debug, PartialEq)]
pub struct Pixel {
    r: u8,
    g: u8,
    b: u8,
}

pub fn compress_64(width: u16, height: u16, pixels: Vec<Pixel>) -> Result<Vec<Pixel>, String> {
    if width * height != pixels.len() as u16 {
        return Err(format!("Number of pixels ({}) not matching width ({}) multiplied by height ({})", pixels.len(), width, height));
    }

    let mut output = vec![Pixel { r: 0, g: 0, b: 0 }; 64];
    for new_y in 0..8 {
        for new_x in 0..8 {
            let mut section = vec![];
            for y in (new_y * height / 8)..((new_y + 1) * height / 8) {
                for x in (new_x * width / 8)..((new_x + 1) * width / 8) {
                    section.push(pixels[(height * y + x) as usize]);
                }
            }
            output[(new_y * 8 + new_x) as usize] = compress(section);
        }
    }

    return Ok(output);
}

pub fn compress(pixels: Vec<Pixel>) -> Pixel {
    if pixels.is_empty() {
        return Pixel { r: 0, g: 0, b: 0 };
    }

    let mut r = 0u32;
    let mut g = 0u32;
    let mut b = 0u32;

    for i in 0..pixels.len() {
        r += pixels[i].r as u32;
        g += pixels[i].g as u32;
        b += pixels[i].b as u32;
    }

    let len = pixels.len() as u32;

    return Pixel {
        r: (r / len) as u8,
        g: (g / len) as u8,
        b: (b / len) as u8,
    };
}

#[cfg(test)]
mod tests {
    // Note this useful idiom: importing names from outer (for mod tests) scope.
    use super::*;

    #[test]
    fn test_compress_64_invalid_input_should_return_error() {
        let result = compress_64(16, 16, vec![]);
        assert!(result.is_err(), "result should be an error, got {:?}", result);
    }

    #[test]
    fn test_compress_64_valid_input_should_compress_picture() {
        // black
        let b = Pixel { r: 0, g: 0, b: 0 };
        // white
        let w = Pixel { r: 255, g: 255, b: 255 };
        // grey
        let g = Pixel { r: 127, g: 127, b: 127 };

        let input = vec![
            b, b, b, b, b, b, b, b, b, b, b, b, b, b, b, b,
            w, w, b, b, w, w, b, b, w, w, b, b, w, w, b, b,
            w, w, w, w, w, w, w, w, w, w, w, w, w, w, w, w,
            w, w, b, b, w, w, b, b, w, w, b, b, w, w, b, b,
            b, b, b, b, b, b, b, b, b, b, b, b, b, b, b, b,
            w, w, b, b, w, w, b, b, w, w, b, b, w, w, b, b,
            w, w, w, w, w, w, w, w, w, w, w, w, w, w, w, w,
            w, w, b, b, w, w, b, b, w, w, b, b, w, w, b, b,
            b, b, b, b, b, b, b, b, b, b, b, b, b, b, b, b,
            w, w, b, b, w, w, b, b, w, w, b, b, w, w, b, b,
            w, w, w, w, w, w, w, w, w, w, w, w, w, w, w, w,
            w, w, b, b, w, w, b, b, w, w, b, b, w, w, b, b,
            b, b, b, b, b, b, b, b, b, b, b, b, b, b, b, b,
            w, w, b, b, w, w, b, b, w, w, b, b, w, w, b, b,
            w, w, w, w, w, w, w, w, w, w, w, w, w, w, w, w,
            w, w, b, b, w, w, b, b, w, w, b, b, w, w, b, b,
        ];

        let expected_output = vec![
            g, b, g, b, g, b, g, b,
            w, g, w, g, w, g, w, g,
            g, b, g, b, g, b, g, b,
            w, g, w, g, w, g, w, g,
            g, b, g, b, g, b, g, b,
            w, g, w, g, w, g, w, g,
            g, b, g, b, g, b, g, b,
            w, g, w, g, w, g, w, g,
        ];

        assert_eq!(compress_64(16, 16, input), Ok(expected_output));
    }

    #[test]
    fn test_compress_no_pixels_should_return_black() {
        assert_eq!(compress(vec![]), Pixel { r: 0, g: 0, b: 0 });
    }

    #[test]
    fn test_compress_monochrome_pixels_return_same_pixel() {
        let count = rand::random::<usize>() % 1023 + 1;
        let pixel = Pixel {
            r: rand::random::<u8>(),
            g: rand::random::<u8>(),
            b: rand::random::<u8>(),
        };
        let pixels = vec![pixel; count];

        assert_eq!(compress(pixels), pixel);
    }

    #[test]
    fn test_compress_shades_of_red_return_shade_of_red() {
        let count = rand::random::<usize>() % 1023 + 1;
        let mut pixels = vec![Pixel { r: 0, g: 0, b: 0 }; count];
        pixels[0].r = 30;
        for i in 1..pixels.len() {
            pixels[i].r = rand::random::<u8>();
        }

        let compressed_pixel = compress(pixels);
        assert!(compressed_pixel.r > 0, "red channel should be greater than zero, got {}", compressed_pixel.r);
        assert_eq!(compressed_pixel.g, 0);
        assert_eq!(compressed_pixel.b, 0);
    }

    #[test]
    fn test_compress_shades_of_green_return_shade_of_green() {
        let count = rand::random::<usize>() % 1023 + 1;
        let mut pixels = vec![Pixel { r: 0, g: 0, b: 0 }; count];
        pixels[0].g = 30;
        for i in 1..pixels.len() {
            pixels[i].g = rand::random::<u8>();
        }

        let compressed_pixel = compress(pixels);
        assert_eq!(compressed_pixel.r, 0);
        assert!(compressed_pixel.g > 0, "green channel should be greater than zero, got {}", compressed_pixel.g);
        assert_eq!(compressed_pixel.b, 0);
    }

    #[test]
    fn test_compress_shades_of_blue_return_shade_of_blue() {
        let count = rand::random::<usize>() % 1023 + 1;
        let mut pixels = vec![Pixel { r: 0, g: 0, b: 0 }; count];
        pixels[0].b = 30;
        for i in 1..pixels.len() {
            pixels[i].b = rand::random::<u8>();
        }

        let compressed_pixel = compress(pixels);
        assert_eq!(compressed_pixel.r, 0);
        assert_eq!(compressed_pixel.g, 0);
        assert!(compressed_pixel.b > 0, "blue channel should be greater than zero, got {}", compressed_pixel.b);
    }

    #[test]
    fn test_compress_return_between_extreme_values() {
        let mut pixels = vec![Pixel { r: 0, g: 0, b: 0 }; 2];

        let mut min_r = u8::MAX;
        let mut max_r = u8::MIN;
        let mut min_g = u8::MAX;
        let mut max_g = u8::MIN;
        let mut min_b = u8::MAX;
        let mut max_b = u8::MIN;

        for i in 0..pixels.len() {
            pixels[i].r = rand::random::<u8>();
            pixels[i].g = rand::random::<u8>();
            pixels[i].b = rand::random::<u8>();

            if pixels[i].r < min_r {
                min_r = pixels[i].r;
            }

            if pixels[i].r > max_r {
                max_r = pixels[i].r;
            }

            if pixels[i].g < min_g {
                min_g = pixels[i].g;
            }

            if pixels[i].g > max_g {
                max_g = pixels[i].g;
            }

            if pixels[i].b < min_b {
                min_b = pixels[i].b;
            }

            if pixels[i].b > max_b {
                max_b = pixels[i].b;
            }
        }

        let compressed_pixel = compress(pixels);
        assert!(min_r <= compressed_pixel.r && max_r >= compressed_pixel.r, "red channel should be in [{}; {}], got {}", min_r, max_r, compressed_pixel.r);
        assert!(min_g <= compressed_pixel.g && max_g >= compressed_pixel.g, "green channel should be in [{}; {}], got {}", min_g, max_g, compressed_pixel.g);
        assert!(min_b <= compressed_pixel.b && max_b >= compressed_pixel.b, "blue channel should be in [{}; {}], got {}", min_b, max_b, compressed_pixel.b);
    }
}
