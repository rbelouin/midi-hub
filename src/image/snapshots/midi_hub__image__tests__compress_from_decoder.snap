---
source: src/image/mod.rs
assertion_line: 133
expression: result

---
Ok(
    [
        Pixel {
            r: 6,
            g: 59,
            b: 51,
        },
        Pixel {
            r: 11,
            g: 82,
            b: 73,
        },
        Pixel {
            r: 19,
            g: 101,
            b: 92,
        },
        Pixel {
            r: 27,
            g: 116,
            b: 106,
        },
        Pixel {
            r: 29,
            g: 122,
            b: 110,
        },
        Pixel {
            r: 26,
            g: 117,
            b: 107,
        },
        Pixel {
            r: 17,
            g: 103,
            b: 95,
        },
        Pixel {
            r: 8,
            g: 89,
            b: 81,
        },
        Pixel {
            r: 9,
            g: 81,
            b: 71,
        },
        Pixel {
            r: 18,
            g: 109,
            b: 99,
        },
        Pixel {
            r: 98,
            g: 101,
            b: 69,
        },
        Pixel {
            r: 156,
            g: 84,
            b: 44,
        },
        Pixel {
            r: 168,
            g: 91,
            b: 43,
        },
        Pixel {
            r: 103,
            g: 112,
            b: 82,
        },
        Pixel {
            r: 30,
            g: 135,
            b: 123,
        },
        Pixel {
            r: 16,
            g: 110,
            b: 100,
        },
        Pixel {
            r: 11,
            g: 101,
            b: 91,
        },
        Pixel {
            r: 52,
            g: 123,
            b: 103,
        },
        Pixel {
            r: 174,
            g: 115,
            b: 81,
        },
        Pixel {
            r: 213,
            g: 191,
            b: 180,
        },
        Pixel {
            r: 208,
            g: 180,
            b: 164,
        },
        Pixel {
            r: 181,
            g: 111,
            b: 67,
        },
        Pixel {
            r: 71,
            g: 133,
            b: 110,
        },
        Pixel {
            r: 24,
            g: 131,
            b: 120,
        },
        Pixel {
            r: 12,
            g: 115,
            b: 105,
        },
        Pixel {
            r: 53,
            g: 132,
            b: 115,
        },
        Pixel {
            r: 193,
            g: 161,
            b: 145,
        },
        Pixel {
            r: 194,
            g: 169,
            b: 163,
        },
        Pixel {
            r: 203,
            g: 179,
            b: 174,
        },
        Pixel {
            r: 191,
            g: 150,
            b: 128,
        },
        Pixel {
            r: 88,
            g: 146,
            b: 121,
        },
        Pixel {
            r: 29,
            g: 143,
            b: 132,
        },
        Pixel {
            r: 12,
            g: 124,
            b: 114,
        },
        Pixel {
            r: 68,
            g: 154,
            b: 141,
        },
        Pixel {
            r: 221,
            g: 201,
            b: 190,
        },
        Pixel {
            r: 209,
            g: 186,
            b: 181,
        },
        Pixel {
            r: 212,
            g: 189,
            b: 186,
        },
        Pixel {
            r: 226,
            g: 207,
            b: 199,
        },
        Pixel {
            r: 95,
            g: 170,
            b: 156,
        },
        Pixel {
            r: 32,
            g: 147,
            b: 137,
        },
        Pixel {
            r: 8,
            g: 123,
            b: 114,
        },
        Pixel {
            r: 47,
            g: 135,
            b: 117,
        },
        Pixel {
            r: 203,
            g: 169,
            b: 147,
        },
        Pixel {
            r: 223,
            g: 203,
            b: 195,
        },
        Pixel {
            r: 223,
            g: 203,
            b: 196,
        },
        Pixel {
            r: 209,
            g: 179,
            b: 158,
        },
        Pixel {
            r: 70,
            g: 145,
            b: 123,
        },
        Pixel {
            r: 26,
            g: 139,
            b: 130,
        },
        Pixel {
            r: 5,
            g: 116,
            b: 108,
        },
        Pixel {
            r: 28,
            g: 139,
            b: 125,
        },
        Pixel {
            r: 148,
            g: 144,
            b: 113,
        },
        Pixel {
            r: 217,
            g: 192,
            b: 180,
        },
        Pixel {
            r: 217,
            g: 191,
            b: 179,
        },
        Pixel {
            r: 174,
            g: 150,
            b: 118,
        },
        Pixel {
            r: 43,
            g: 139,
            b: 123,
        },
        Pixel {
            r: 18,
            g: 125,
            b: 116,
        },
        Pixel {
            r: 3,
            g: 103,
            b: 96,
        },
        Pixel {
            r: 6,
            g: 121,
            b: 113,
        },
        Pixel {
            r: 13,
            g: 140,
            b: 130,
        },
        Pixel {
            r: 30,
            g: 147,
            b: 136,
        },
        Pixel {
            r: 34,
            g: 147,
            b: 135,
        },
        Pixel {
            r: 25,
            g: 140,
            b: 129,
        },
        Pixel {
            r: 16,
            g: 125,
            b: 117,
        },
        Pixel {
            r: 9,
            g: 107,
            b: 100,
        },
    ],
)
